guide/src/ecosystem/async-await.md

```diff
@@ -154,6 +341,360 @@ async fn main() -> PyResult<()> {
+```
+
+#### Using `uvloop` in Rust Applications
+
+Using `uvloop` in Rust applications is a bit trickier, but it's still possible
+with relatively few modifications.
+
+> Unfortunately, we can't make use of the `#[pyo3_asyncio::<runtime>::main]` attribute with non-standard event loops. This is because the `#[pyo3_asyncio::<runtime>::main]` proc macro has to interact with the Python
+event loop before we can install the `uvloop` policy.
+
+```toml
+[dependencies]
+async-std = "1.9"
+pyo3 = "0.14"
+pyo3-asyncio = { version = "0.14", features = ["async-std-runtime"] }
+```
+
+```rust
+//! main.rs
+
+use pyo3::{prelude::*, types::PyType};
+
+fn main() -> PyResult<()> {
+    pyo3::prepare_freethreaded_python();
+
+    Python::with_gil(|py| {
+        let uvloop = py.import("uvloop")?;
+        uvloop.call_method0("install")?;
+
+        // store a reference for the assertion
+        let uvloop = PyObject::from(uvloop);
+
+        pyo3_asyncio::async_std::run(py, async move {
+            // verify that we are on a uvloop.Loop
+            Python::with_gil(|py| -> PyResult<()> {
+                assert!(uvloop
+                    .as_ref(py)
+                    .getattr("Loop")?
+                    .downcast::<PyType>()
+                    .unwrap()
+                    .is_instance(pyo3_asyncio::async_std::get_current_loop(py)?)?);
+                Ok(())
+            })?;
+
+            async_std::task::sleep(std::time::Duration::from_secs(1)).await;
+
+            Ok(())
+        })
+    })
+}
+```
+
+### Event Loop References and Thread-awareness
+
+One problem that arises when interacting with Python's asyncio library is that the functions we use to get a reference to the Python event loop can only be called in certain contexts. Since PyO3 Asyncio needs to interact with Python's event loop during conversions, the context of these conversions can matter a lot. 
+
+> The core conversions we've mentioned so far in this guide should insulate you from these concerns in most cases, but in the event that they don't, this section should provide you with the information you need to solve these problems.
+
+#### The Main Dilemma
+
+Python programs can have many independent event loop instances throughout the lifetime of the application (`asyncio.run` for example creates its own event loop each time it's called for instance), and they can even run concurrent with other event loops. For this reason, the most correct method of obtaining a reference to the Python event loop is via `asyncio.get_running_loop`.
+
+`asyncio.get_running_loop` returns the event loop associated with the current OS thread. It can be used inside Python coroutines to spawn concurrent tasks, interact with timers, or in our case signal between Rust and Python. This is all well and good when we are operating on a Python thread, but since Rust threads are not associated with a Python event loop, `asyncio.get_running_loop` will fail when called on a Rust runtime.
+
+#### The Solution
```

Comment:
This section is quite long, and goes through two "not quite correct" examples before eventually reaching a set of recommended functions to use.

I feel like users might get lost after copy-pasting either the first or second example and never see the third "recommended" method.

How do you feel about turning this section on its head a bit? Put the recommended APIs and example first, and then offer up a subsection for power users to see `into_future_with_loop` / `scope` etc.

I think that'd let most users see "the solution" immediately, and give advanced users extra material to dive into the nuts and bolts. It might even be nice to put the other "not quite correct" examples somewhere in the PyO3 Asyncio library docs and just link to them from here.

Reply:
I see what you mean. The general idea I had for this section was to explain the problems, then provide the intuition behind the new functions in `v0.14` by building to the final recommended solution.

I put this section below the quickstart and general usage because it was intended to get more into the weeds _after_ showing the simpler usage. Maybe this shouldn't even go in the guide though. It might be better served in the API docs with links in the guide. What do you think?

Reply:
That seems like a good idea to me. In general I'd prefer to keep this document full of examples of how to get started with PyO3 Asyncio, and let really rich documentation like this for low-level APIs live as close to those APIs as possible (as long as there's ways for users to discover them in the first place).